src/Orckestra.Composer.Cart/Providers/Order/EditingOrderProvider.cs

```diff
@@ -53,6 +54,67 @@ public virtual async Task<bool> CanEdit(Overture.ServiceModel.Orders.Order order
             return isOrderEditable;
         }
 
+        public virtual async Task<bool> CanCancel(Overture.ServiceModel.Orders.Order order)
+        {
+            if (order?.Cart?.Shipments == null
+                || OrderHistoryConfiguration.CompletedOrderStatuses.Contains(order.OrderStatus)
+                || string.IsNullOrWhiteSpace(order.OrderStatus))
+            {
+                return false;
+            }
+            var orderFulfillmentState = await OrderRepository.GetOrderFulfillmentStateAsync(new GetOrderFulfillmentStateParam
+            {
+                OrderId = order.Id,
+                ScopeId = order.ScopeId
+            }).ConfigureAwait(false);
+
+            if (orderFulfillmentState.IsCancelable
+                && !orderFulfillmentState.IsProcessing
+                && !orderFulfillmentState.Status.Equals(Constants.OrderStatus.Canceled,
+                    StringComparison.InvariantCultureIgnoreCase)
+                && !orderFulfillmentState.Status.Equals("Completed", StringComparison.InvariantCultureIgnoreCase)
+                && orderFulfillmentState.ShipmentFulfillmentStates.All(item => item.AllowedStatusChanges.Contains(Constants.OrderStatus.Canceled)))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        public virtual async Task<bool> PendingCancel(Overture.ServiceModel.Orders.Order order)
+        {
+            if (order.OrderStatus.Equals(Constants.OrderStatus.Canceled, StringComparison.InvariantCultureIgnoreCase)) return false;
+
+            var orderFulfillmentState = await OrderRepository.GetOrderFulfillmentStateAsync(
+                new GetOrderFulfillmentStateParam
+                {
+                    OrderId = order.Id,
+                    ScopeId = order.ScopeId
+                }).ConfigureAwait(false);
+
+            return !orderFulfillmentState.IsCancelable;
+        }
+
+        public async Task CancelOrder(Overture.ServiceModel.Orders.Order order, string scope)
+        {
+            var isOrderCancelable = await CanCancel(order).ConfigureAwait(false);
+            if (!isOrderCancelable) throw new InvalidOperationException($"Order {order.Id} cann't be canceled");
+            
+            var shipments = order.Cart?.Shipments;
+            
```

Comment:
please use ask.WhenAll(shipmentsTasks).ConfigureAwait(false) instead of foreach

shipmentsTasks = order.Cart?.Shipments?.Select(shipment=> await OrderRepository.ChangeShipmentStatusAsync(,...)